src/cloudWatchLogs/commands/searchLogGroup.ts

```diff
@@ -149,18 +146,95 @@ async function logGroupsToArray(logGroups: AsyncIterableIterator<CloudWatchLogs.
     return logGroupsArray
 }
 
-export function createFilterpatternPrompter(logGroupName: string, isFirst: boolean): InputBoxPrompter {
+/**
+ * HACK: this subclass overrides promptUser() so that we can validate the
+ * search pattern against the service and if it fails, keep the prompt displayed.
+ *
+ * This is necessary until vscode's inputbox.onDidAccept() awaits async callbacks:
+ *    - https://github.com/aws/aws-toolkit-vscode/pull/3114#discussion_r1085484630
+ *    - https://github.com/microsoft/vscode/blob/78947444843f4ebb094e5ab4288360010a293463/extensions/git-base/src/remoteSource.ts#L13
+ *    - https://github.com/microsoft/vscode/blob/78947444843f4ebb094e5ab4288360010a293463/src/vs/base/browser/ui/inputbox/inputBox.ts#L511
+ */
+export class SearchPatternPrompter extends InputBoxPrompter {
+    constructor(
+        public logGroup: CloudWatchLogsGroupInfo,
+        public logParams: CloudWatchLogsParameters,
+        /** HACK: also maintain ad-hoc state because `wizardState` is not mutable. */
+        public readonly retryState: any,
+        public readonly inputBox: InputBox,
+        protected readonly options: ExtendedInputBoxOptions = {}
+    ) {
+        super(inputBox, options)
+
+        // HACK: restore our own ad-hoc state.
+        if (retryState.searchPattern) {
+            this.recentItem = retryState.searchPattern
+        }
+        this.inputBox.validationMessage = retryState.validationMessage ? retryState.validationMessage : undefined
+    }
+
+    protected async promptUser(): Promise<PromptResult<string>> {
+        const rv = await super.promptUser()
+        this.inputBox.busy = true
+
+        const validationResult = await this.validateSearchPattern(this.inputBox.value, true)
+        try {
+            // HACK: maintain our own state and restore it.
+            this.retryState.searchPattern = rv?.toString()
+            if (validationResult !== undefined) {
+                // Doesn't work because state is recreated on WIZARD_RETRY?
+                // this.inputBox.validationMessage = validationResult
+                // HACK: maintain our own state and restore it.
+                this.retryState.validationMessage = validationResult
+
+                if (rv) {
+                    // Doesn't work because state is recreated on WIZARD_RETRY?
+                    // this.recentItem = rv.toString()
```

Comment:
Probably need to remove `recentItem` from the interface as it can be confusing. The wizard control flow logic will override your assignment.